stdlib/public/core/Integers.swift

```diff
@@ -2691,1292 +2659,942 @@ extension FixedWidthInteger {
-  @_transparent
-  public func ${x.name}ReportingOverflow(
-    ${x.firstArg} other: ${Self}
-  ) -> (partialValue: ${Self}, overflow: Bool) {
-%         if x.kind == '/':
-    // No LLVM primitives for checking overflow of division operations, so we
-    // check manually.
-    if _slowPath(other == (0 as ${Self})) {
-      return (partialValue: self, overflow: true)
+  /// Creates a new instance from the bit pattern of the given instance by
+  /// truncating or sign-extending if needed to fit this type.
+  ///
+  /// When the bit width of `T` (the type of `source`) is equal to or greater
+  /// than this type's bit width, the result is the truncated
+  /// least-significant bits of `source`. For example, when converting a
+  /// 16-bit value to an 8-bit type, only the lower 8 bits of `source` are
+  /// used.
+  ///
+  ///     let p: Int16 = -500
+  ///     // 'p' has a binary representation of 11111110_00001100
+  ///     let q = Int8(truncatingIfNeeded: p)
+  ///     // q == 12
+  ///     // 'q' has a binary representation of 00001100
+  ///
+  /// When the bit width of `T` is less than this type's bit width, the result
+  /// is *sign-extended* to fill the remaining bits. That is, if `source` is
+  /// negative, the result is padded with ones; otherwise, the result is
+  /// padded with zeros.
+  ///
+  ///     let u: Int8 = 21
+  ///     // 'u' has a binary representation of 00010101
+  ///     let v = Int16(truncatingIfNeeded: u)
+  ///     // v == 21
+  ///     // 'v' has a binary representation of 00000000_00010101
+  ///
+  ///     let w: Int8 = -21
+  ///     // 'w' has a binary representation of 11101011
+  ///     let x = Int16(truncatingIfNeeded: w)
+  ///     // x == -21
+  ///     // 'x' has a binary representation of 11111111_11101011
+  ///     let y = UInt16(truncatingIfNeeded: w)
+  ///     // y == 65515
+  ///     // 'y' has a binary representation of 11111111_11101011
+  ///
+  /// - Parameter source: An integer to convert to this type.
+  @inlinable // FIXME(sil-serialize-all)
+  @inline(__always)
+  public init<T : BinaryInteger>(truncatingIfNeeded source: T) {
+    if Self.bitWidth <= Int.bitWidth {
+      self = Self.init(_truncatingBits: source._lowWord)
     }
-%           if signed:
-    if _slowPath(self == ${Self}.min && other == (-1 as ${Self})) {
-%             partialValue = 'self' if x.operator == '/' else '0'
-      return (partialValue: ${partialValue}, overflow: true)
+    else {
+      let neg = source < (0 as T)
+      var result: Self = neg ? ~0 : 0
+      var shift: Self = 0
+      let width = Self(_truncatingBits: Self.bitWidth._lowWord)
+      for word in source.words {
+        guard shift < width else { break }
+        // Masking shift is OK here because we have already ensured
+        // that shift < Self.bitWidth. Not masking results in
+        // infinite recursion.
+        result ^= Self(_truncatingBits: neg ? ~word : word) &<< shift
+        shift += Self(Int.bitWidth)
```

Comment:
I'm a little worried about this line. It used to be a literal instead of a `Self(Int.bitWidth)`...